src/lib/crypto/kimchi_backend/gadgets/foreign_field.ml

```diff
@@ -768,142 +822,52 @@ let sum_setup (type f) (module Circuit : Snark_intf.Run with type field = f)
-  Field.Assert.equal ovf Field.one ;
-
-  (* Check that the highest limb of right input is 2^88*)
-  let two_to_88 = two_to_limb_field (module Circuit) in
-  Field.Assert.equal (Field.constant two_to_88) offset2 ;
-
-  (* Add external check for multi range check *)
-  External_checks.append_multi_range_check external_checks
-  @@ Element.Standard.to_limbs bound ;
-
-  (* Return the bound value *)
-  bound
-
-(* Gadget to constrain external checks using supplied modulus *)
-let constrain_external_checks (type field)
-    (module Circuit : Snark_intf.Run with type field = field)
-    (external_checks : field External_checks.t) (modulus : field standard_limbs)
-    =
-  (* 1) Add gates for external bound additions.
-   *    Note: internally this also adds multi-range-checks for the
-   *    computed bound to the external_checks.multi-ranges, which
-   *    are then constrainted in (2)
-   *)
-  List.iter external_checks.bounds ~f:(fun value ->
-      let _bound =
-        valid_element
-          (module Circuit)
-          external_checks
-          (Element.Standard.of_limbs value)
-          modulus
-      in
-      () ) ;
-
-  (* 2) Add gates for external multi-range-checks *)
-  List.iter external_checks.multi_ranges ~f:(fun multi_range ->
-      let v0, v1, v2 = multi_range in
-      Range_check.multi (module Circuit) v0 v1 v2 ;
-      () ) ;
-
-  (* 3) Add gates for external compact-multi-range-checks *)
-  List.iter external_checks.compact_multi_ranges ~f:(fun compact_multi_range ->
-      let v01, v2 = compact_multi_range in
-      Range_check.compact_multi (module Circuit) v01 v2 ;
-      () )
-
-(* FOREIGN FIELD ADDITION CHAIN GADGET *)
-
 (** Gadget for a chain of foreign field sums (additions or subtractions)
  *
  *    Inputs:
  *      inputs                := All the inputs to the chain of sums
- *      operations            := List of operation modes Add or Sub indicating whether th
+ *      operations            := List of operation modes Add or Sub indicating whether the
  *                               corresponding addition is a subtraction
- *      foreign_field_modulus := The modulus of the foreign field (all the same)
+ *      foreign_field_modulus := Foreign field modulus
  *
  *    Outputs:
- *      Inserts the gates (described below) into the circuit
+ *      Inserts ForeignFieldAdd gate into the circuit
  *      Returns the final result of the chain of sums
  *
- *    For n+1 inputs, the gadget creates n foreign field addition gates, followed by a final
- *    foreign field addition gate for the bound check (i.e. valid_element check). For this, a
- *    an additional multi range check must also be performed.
- *    By default, the range check takes place right after the final Raw row.
+ *    For n+1 inputs, the gadget creates n foreign field addition gates.
+ *
+ * TODO:
+ *    Understand if concatenating sums is possible with input limbs < 2^88 with chunking
```

Comment:
Here's my analysis @jspada: I don't think there is any restriction about how many sums we can concatenate.

Each ffadd gate individually proves that $a + b = qf + r$ as integers, for some $q \in \{ 0, 1 \}$.

So if we concatenate $n$ additions to form the sum $a_0 + \ldots + a_n$ with output $r$, then you can follow this logic to prove that, as integers,

$$
a_0 + \ldots + a_n = (q_1 + \ldots + q_n) f + r
$$

Since each $q_i$ is 0 or 1, the sum of $q_i$s is a number between $0$ and $n$ (inclusive).

The only restriction is that as we add more summands, a full bounds check on the summands tells us less and less about the size of $r$. In particular, $r$ could be the unreduced result of all those additions and be comparable in size to $np$ (with the restriction that $r < 2^{3\ell}$ which we know from the range checks we need on the limbs of $r$).

There isn't any possible soundness issue